app_dart/lib/src/request_handling/cache_request_handler.dart

```diff
@@ -50,33 +50,105 @@ class CacheRequestHandler<T extends Body> extends RequestHandler<T> {
     final responseKey = '${request!.uri.path}:${request!.uri.query}';
 
     if (request!.uri.queryParameters[flushCacheQueryParam] == 'true') {
-      await cache.purge(responseSubcacheName, responseKey);
+      await _cache.purge(responseSubcacheName, responseKey);
     }
 
-    final cachedResponse = await cache.getOrCreateWithLocking(
+    final cachedBytes = await _cache.getOrCreateWithLocking(
       responseSubcacheName,
       responseKey,
-      createFn: () => getBodyBytesFromDelegate(delegate),
-      ttl: ttl,
+      createFn: () async {
+        // TODO(matanlurey): Evaluate if 5XX errors should not be cached.
+        final response = await _createCachedResponse(_delegate);
+        return response.toBytes();
+      },
+      ttl: _ttl,
+    );
+
+    final cachedResponse = _CachedHttpResponse.fromBytes(
+      cachedBytes!,
+      debugName: responseKey,
     );
 
-    return Body.forStream(Stream<Uint8List?>.value(cachedResponse)) as T;
+    response!
+      ..statusCode = cachedResponse.statusCode
+      ..reasonPhrase = cachedResponse.reason;
+
+    return Body.forStream(Stream<Uint8List?>.value(cachedResponse.body)) as T;
   }
 
   /// Get a Uint8List that contains the bytes of the response from [delegate]
-  /// so it can be stored in [cache].
-  Future<Uint8List> getBodyBytesFromDelegate(RequestHandler<T> delegate) async {
-    final Body body = await delegate.get();
-
-    // Body only offers getting a Stream<Uint8List> since it just sends
-    // the data out usually to a client. In this case, we want to store
-    // the bytes in the cache which requires several conversions to get a
-    // Uint8List that contains the bytes of the response.
-    final rawBytes =
-        await body
-            .serialize()
-            .expand<int>((Uint8List? chunk) => chunk!)
-            .toList();
-    return Uint8List.fromList(rawBytes);
+  /// so it can be stored in [_cache].
+
+  /// Invokes [delegate.get], and returns the result as a [_CachedHttpResponse].
+  Future<_CachedHttpResponse> _createCachedResponse(
+    RequestHandler<T> delegate,
+  ) async {
+    final body = await delegate.get();
+    final response = this.response!;
+    final builder = BytesBuilder();
+    await body.serialize().forEach(builder.add);
+    return _CachedHttpResponse._(
+      response.statusCode,
+      response.reasonPhrase,
+      builder.toBytes(),
+    );
+  }
+}
+
+@immutable
+final class _CachedHttpResponse {
+  static const _magic4Bytes = 0xFA_CE_FE_ED;
+  static final _magic4Uint8List =
+      Uint8List(4)
+        ..[0] = 0xFA
+        ..[1] = 0xCE
+        ..[2] = 0xFE
+        ..[3] = 0xED;
+
+  factory _CachedHttpResponse.fromBytes(
+    Uint8List bytes, {
+    required String debugName,
+  }) {
+    if (bytes.length < _magic4Uint8List.length ||
+        bytes.buffer.asByteData().getUint32(0) != _magic4Bytes) {
```

Comment:
```suggestion
        bytes.buffer.asByteData().getUint32(0, Endian.host) != _magic4Bytes) {
```

Reply:
You're right. Passing Endian.big would be explicit.

Or we could just compare the first 4 bytes are the same as _magic4Uint8List